Finished jobs have to send a status email to the mailing list (#45)

cosmin-malutan opened the pull request:
This is for issue #45.
I configured the jobs to send emails, similar as we do on mozmill-ci. Make different reg-ex for the mail subject, and I tested it on staging node, because from local node it can't send emails through mozilla smtp server (smtp.mozilla.org)

cosmin-malutan committed:
Finished jobs have to send a status email to the mailing list (#45)

cosmin-malutan commented:
@davehunt can you review this?

davehunt commented:
I suspect the e-mail notifications will want to be enabled by the jenkins patch so that we can have different settings for each environment. Also, please provide an example console output and e-mail example. If you want to test this locally you could use the technique described here: https://docs.djangoproject.com/en/dev/topics/email/#configuring-email-for-development

cosmin-malutan commented:
Thanks @davehunt I already tested this on a staging node with real mozilla smtp.mozilla.org service. 

Here is an email I've got, the log is a common log we have for tps jobs on [TPS_CI](http://tps-ci-production.qa.scl3.mozilla.com:8080/)

```
Return-Path: <tps-reports@mozilla.org>
X-Original-To: cosmin.malutan@softvision.ro
Delivered-To: cosmin.malutan@softvision.ro
Received: from svmail.softvision.ro (localhost [127.0.0.1])
    by svmail.softvision.ro (Postfix) with ESMTP id C4C3E63F
    for <cosmin.malutan@softvision.ro>; Wed, 20 Aug 2014 09:59:22 +0300 (EEST)
Received: from smtp.mozilla.org (mx2.corp.phx1.mozilla.com [63.245.216.70])
    by svmail.softvision.ro (Postfix) with ESMTPS id 4EF5A615
    for <cosmin.malutan@softvision.ro>; Wed, 20 Aug 2014 09:59:21 +0300 (EEST)
Received: from mm-ub-1310-64.qa.scl3.mozilla.com (nat-qa.scl3.mozilla.com [63.245.214.168])
    by mx2.mail.corp.phx1.mozilla.com (Postfix) with ESMTP id 52334F2045
    for <cosmin.malutan@softvision.ro>; Tue, 19 Aug 2014 23:59:14 -0700 (PDT)
Date: Tue, 19 Aug 2014 23:59:13 -0700 (PDT)
From: tps-reports@mozilla.org
To: cosmin.malutan@softvision.ro
Message-ID: <1420854767.17.1408517954015.JavaMail.mozauto@mm-ub-1310-64.qa.scl3.mozilla.com>
In-Reply-To: <1338317678.15.1408517341974.JavaMail.mozauto@mm-ub-1310-64.qa.scl3.mozilla.com>
References: <1338317678.15.1408517341974.JavaMail.mozauto@mm-ub-1310-64.qa.scl3.mozilla.com>
Subject: [tps-reports][Still Failing] mozilla-central_fx-account: Firefox
 version: 33.0a1 on  (20140709131331)
MIME-Version: 1.0
Content-Type: multipart/mixed; 
    boundary="----=_Part_16_1443823908.1408517953947"
X-Jenkins-Job: mozilla-central_fx-account
X-Jenkins-Result: FAILURE
List-ID: Notifications from the TPS-CI system <tps-reports.mozilla.org>
X-Virus-Scanned: clamav-milter 0.97.8 at svmail
X-Virus-Status: Clean

------=_Part_16_1443823908.1408517953947
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 7bit

TPS mozilla-central_fx-account testrun for Firefox version: 33.0a1 on  (20140709131331) completed with failures.

Test Summary

TEST-PASS | test_sync.js | 
TEST-PASS | test_prefs.js | 
TEST-UNEXPECTED-FAIL | test_tabs.js | [phase2] Exception caught: ASSERTION FAILED! error locating tab No traceback available
TEST-PASS | test_passwords.js | 
TEST-PASS | test_history.js | 
TEST-PASS | test_formdata.js | 
TEST-PASS | test_bug530717.js | 
TEST-PASS | test_bug531489.js | 
TEST-PASS | test_bug538298.js | 
TEST-PASS | test_bug556509.js | 
TEST-PASS | test_bug562515.js | 
TEST-PASS | test_bug563989.js | 
TEST-PASS | test_bug535326.js | 
TEST-PASS | test_bug501528.js | 
TEST-PASS | test_bug575423.js | 
TEST-PASS | test_bug546807.js | 
TEST-PASS | test_history_collision.js | 
TEST-PASS | test_privbrw_formdata.js | 
TEST-PASS | test_privbrw_passwords.js | 
TEST-PASS | test_privbrw_tabs.js | 
TEST-PASS | test_bookmarks_in_same_named_folder.js | 
TEST-PASS | test_client_wipe.js | 
TEST-PASS | test_special_tabs.js | 
TEST-PASS | test_addon_sanity.js | 
TEST-PASS | test_addon_restartless_xpi.js | 
TEST-PASS | test_addon_nonrestartless_xpi.js | 
TEST-PASS | test_addon_reconciling.js | 
TEST-PASS | test_addon_wipe.js | 
GET https://api.accounts.firefox.com/__heartbeat__


View the build in Jenkins:
http://localhost:8080/job/mozilla-central_fx-account/40/
------=_Part_16_1443823908.1408517953947
Content-Type: application/octet-stream; name=build.log
Content-Transfer-Encoding: base64
Content-Disposition: attachment; filename=build.log

```

davehunt commented:
It looks like `${NODE_NAME}` is not expanded in the e-mail you received.

davehunt commented on jenkins-master/jobs/mozilla-aurora_fx-account/config.xml:
```diff
@@ -67,7 +67,73 @@
+        </hudson.plugins.emailext.plugins.trigger.FailureTrigger>
+      </configuredTriggers>
+      <contentType>default</contentType>
+      <defaultSubject>[tps-reports][${BUILD_STATUS}] ${PROJECT_NAME}: ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: )(Firefox version: .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTrun ...
```
As this is from `tps-reports@mozilla.org` do we need `[tps-reports]` in the subject?

davehunt replied:
I think it would look better to have `Firefox 33.0a1` rather than `Firefox version: 33.0a1` in the subject.

cosmin-malutan replied:
I will change that. :)

davehunt replied:
This states the build completed with failures, which is not accurate for an aborted build. In Mozmill CI we have a custom e-mail content for this trigger.

davehunt replied:
I still see the `[tps-reports]` prefix in the subject, did we decide to keep this?

davehunt commented on jenkins-master/jobs/mozilla-aurora_fx-account/config.xml:
```diff
@@ -67,7 +67,73 @@
+      <defaultSubject>[tps-reports][${BUILD_STATUS}] ${PROJECT_NAME}: ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: )(Firefox version: .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTrun ...
+      <defaultContent>TPS ${PROJECT_NAME} testrun for ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: )(Firefox version: .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTruncatedLines=false ...
+
+${BUILD_LOG_REGEX, regex=&quot;^(Test Summary)$&quot;, showTruncatedLines=false, substText=&quot;$1&quot;, linesBefore=0,linesAfter=29}
```
Including the test summary is a nice idea, but is this just including 29 lines from the start of the summary? What if we added/removed tests?

cosmin-malutan commented:
@davehunt I made the changes, I made a different template for aborted jobs, for test failures I added only the failures and not the whole summary, this adds the failures no matter how many failures we have. I'll forward you an example. ;) 
Thanks

davehunt commented on config/production/jenkins.patch:
```diff
@@ -112,7 +112,7 @@ index 66405c8..fbb7d90 100644
            </default>
            <int>1</int>
-           <string>$NOTIFICATION_ADDRESS</string>
+           <string>NOTIFICATION_ADDRESS</string>
```
This bit shouldn't be in the patch, it looks like it may have been a typo initially.

davehunt replied:
My mistake, this is fine. Reading diffs of patches is hard. :)

davehunt commented on jenkins-master/jobs/mozilla-aurora_fx-account/config.xml:
```diff
@@ -67,7 +67,76 @@
+      <defaultSubject>[tps-reports][${BUILD_STATUS}] ${PROJECT_NAME}: ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: )(Firefox version: .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTrun ...
+      <defaultContent>TPS ${PROJECT_NAME} testrun for Firefox ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox version:)( .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTruncatedLin ...
+
+${BUILD_LOG_REGEX, regex=&quot;^(TEST-UNEXPECTED-FAIL.*)$&quot;, maxMatches=100, showTruncatedLines=false, substText=&quot;$1&quot;}
```
I like this, but we should have some lead text on the line before the failures. Something like `Test failures:`

davehunt commented on jenkins-master/jobs/mozilla-central_fx-account/config.xml:
```diff
@@ -61,13 +61,82 @@
     </hudson.plugins.copyartifact.CopyArtifact>
     <hudson.plugins.xshell.XShellBuilder plugin="xshell@0.9">
-      <commandLine>python trigger.py --urltest $URL_TESTS --urlbuild $URL_BUILD --debug $DEBUG</commandLine>
+      <commandLine>python print.py</commandLine>
```
Is this an intentional change?

cosmin-malutan replied:
No, I tested with that, to avoid running complete testruns on every change. Sorry again for this too.

davehunt commented on jenkins-master/jobs/release-mozilla-release_fx-account/config.xml:
```diff
@@ -67,7 +67,76 @@
+        </hudson.plugins.emailext.plugins.trigger.FailureTrigger>
+      </configuredTriggers>
+      <contentType>default</contentType>
+      <defaultSubject>[tps-reports][${BUILD_STATUS}] ${PROJECT_NAME}: ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: )(Firefox version: .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTrun ...
```
The subjects are inconsistent, sometimes we have the `[tps-reports]` prefix and sometimes not.

cosmin-malutan replied:
I missed to backport the change in all jobs, sorry.

cosmin-malutan commented:
@davehunt I made the changes, sorry for inconsistency.

davehunt commented:
This looks okay to me. Let's have @whimboo perform the final review though.

cosmin-malutan commented:
Sure, thanks a lot Dave :)

whimboo commented on jenkins-master/hudson.tasks.Mailer.xml:
```diff
@@ -1,7 +1,9 @@
 <?xml version='1.0' encoding='UTF-8'?>
 <hudson.tasks.Mailer_-DescriptorImpl plugin="mailer@1.6">
   <defaultSuffix>@mozilla.com</defaultSuffix>
+  <adminAddress>tps-reports@mozilla.com</adminAddress>
```
This is the wrong email address, which even doesn't exist. We could use automation@mozilla.com here.

whimboo commented on jenkins-master/jobs/mozilla-aurora_fx-account/config.xml:
```diff
@@ -67,7 +67,77 @@
+          <email>
+            <recipientList></recipientList>
+            <subject>$PROJECT_DEFAULT_SUBJECT</subject>
+            <body>TPS ${PROJECT_NAME} testrun for Firefox ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox version:)( .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTruncatedLines=f ...
```
What we need here is the tests status as overview directly in the email. We don't want to send out emails with only the link to the jenkins job, and attachments. It has to be quickly readable by the person, who checks results.

Also make sure if you test this, you do NOT use the official mailing list. This is disturbing.

cosmin-malutan replied:
This it will take the default email subject and status form:
https://github.com/mozilla/coversheet/pull/50/files#diff-c567af5a6d17cb8b33d678dff0d777a1R125 
I'll forward you some of emails from the time I made the pr, beside that I'll update the email address.

whimboo replied:
Those emails are looking fine to me regarding the shown content. Thanks!

cosmin-malutan commented:
@whimboo I updated the email addresses, but as a fact they do exists, we used them in the old coversheet instance. Regarding the emails statuses I forwarded you an email as an example, I used regex to give the status in the subject and the testes which failed in the email content, so I would say the second request is covered.

cosmin-malutan commented:
I added email templates for successful jobs, and I changed back the notification addresses to tps-reports@mozilla.com and the admin address I left it automation@mozilla.com. I forwarded all four types of emails to @whimboo and @edmoz, and I'm waiting for feedback.

whimboo commented on config/production/jenkins.patch:
```diff
@@ -112,9 +112,9 @@ index 66405c8..fbb7d90 100644
+           <string>NOTIFICATION_ADDRESS</string>
 -          <string></string>
-+          <string>tps-reports@mozilla.org</string>
++          <string>tps-reports.mozilla.com</string>
```
What? Why did you do this change? Please check your description to the mailing list and you will see it is `.org` and not `.com`.

whimboo replied:
Also why did you replace the `@` with `.`?

cosmin-malutan replied:
No idea. I fixed it.

whimboo commented on config/production/jenkins.patch:
```diff
@@ -22,8 +22,8 @@ index 66405c8..fbb7d90 100644
-+          <onlineAddresses>tps-reports@mozilla.org</onlineAddresses>
-+          <offlineAddresses>tps-reports@mozilla.org</offlineAddresses>
++          <onlineAddresses>tps-reports.mozilla.com</onlineAddresses>
++          <offlineAddresses>tps-reports.mozilla.com</offlineAddresses>
```
the watcher plugin still cannot use variables? I really would like to see that $NOTIFICATION_ADDRESS is used everywhere.

whimboo replied:
Oh, I see https://issues.jenkins-ci.org/browse/JENKINS-20536 which is still unresolved.

whimboo commented on jenkins-master/jobs/mozilla-aurora_fx-account/config.xml:
```diff
@@ -67,7 +67,96 @@
+          <email>
+            <recipientList></recipientList>
+            <subject>$PROJECT_DEFAULT_SUBJECT</subject>
+            <body>TPS ${PROJECT_NAME} testrun for Firefox ${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox version:)( .*)$&quot;, maxMatches=1, showTruncatedLines=false, substText=&quot;$1&quot;, addNewline=false, defaultValue=&quot;%application%&quot;} on ${NODE_NAME} (${BUILD_LOG_REGEX, regex=&quot;^(?:CROSSWEAVE INFO: Firefox buildid: )(.*)$&quot;, maxMatches=1, showTruncatedLines=f ...
```
nit: 'compleated'. Also there is an extra blank between Firefox and the version.

whimboo commented:
Beside the issue as mentioned above all should be fine now.

cosmin-malutan commented:
Fixed the nit and rebased. Thanks for review @whimboo.

whimboo commented:
Merged as https://github.com/mozilla/coversheet/commit/ca355bf2930a1baa122af4b20834c911649576f3
